### ISSUE TITLE
Configuration duplication in preCICE and adapters: mesh and read/write data

### ISSUE THREAD

**Issue by MakisH** [precice/precice#474]
# Problem

At the moment, there are options that need to be set both in the preCICE configuration file and in every adapter's configuration file, in exactly the same way.

This duplication can lead to errors, requires more checks from preCICE/adapters and reduces usability. Even if the user should understand what she is doing, she should not need to adapt to our technical complexity, but we should make the system more tight and less complex.

# An example

Let's look at an example: the OpenFOAM-OpenFOAM CHT tutorial with nearest-projection mapping. In the Fluid participant, we set:

https://github.com/precice/openfoam-adapter/blob/88dfbbae21a0dbdc9f21357a2a1d5d1954646a61/tutorials/CHT/flow-over-plate/buoyantPimpleFoam-laplacianFoam_nearest-projection/Fluid/precice-adapter-config.yml#L5-L17

The `read-data` and `write-data` are also already known to preCICE, as they are set in the configuration file:

https://github.com/precice/openfoam-adapter/blob/88dfbbae21a0dbdc9f21357a2a1d5d1954646a61/tutorials/CHT/flow-over-plate/buoyantPimpleFoam-laplacianFoam_nearest-projection/precice-config_serial.xml#L31-L36

Ideally, we could directly get groups of parameters, such as: {mesh, read/write, data name}. Although we still need to associate each solver interface with a part of the preCICE configuration file. As soon as this association is established, there should be no duplication.

Moreover, the `provideMeshConnectivity` boolean is a technical thing that the user should not need to know about, as preCICE knows if it needs it. This is described in issue #467. As I wrote on the same issue:
> I actually like this idea also for consistency with the high-level API idea. For example, the user does not need to configure an adapter to keep checkpoints when asking for implicit coupling. It's rather the adapter that asks preCICE if this is needed. Why not do the same here?

This is also related to #311, in the sense that we duplicate information that preCICE already knows.

# Potential solution

Implent additional methods in the API, which return what a participant should read/write on which mesh and which interface.

**Comment by uekerman** [comment]
I agree that we could simplify the life of our users here. 

For simple cases, such as `provideMeshConnectivity`, there are straight-forward solutions that are certainly worth the effort #467. 

For handling coupling meshes and data, things are more complicated. 

Back in the old days, Bernhard had already kind of a solution to this:
* In `Constants.hpp` we define default names for data types, such as `Displacements`, `Forces`, ... and also a default name for the preCICE config. 
* In the API, we use `hasMesh` and `hasData` to check whether the default fields are defined in the preCICE config and create the necessary fields in the adapter (we still have these functions in the API). 

The downside of this solution is that it does not work for multiple coupling interfaces (at least not without additional hacking) and that it is bound to a certain type of coupling physics -- sth we want to keep out of preCICE for a clean separation adapter <-> library. 

Let's look at an example. 

(1) Let's say you want to couple two OpenFOAM instances. Is there a way that the adapter can find out whether the participant name is `OpenFOAM1` or `OpenFOAM2` (or `Inner-Fluid`, `Outer-Fluid`)? Everything else depends on this first step.

(2) Let's say, we still pass the participant name through the adapter config. Then, we could have two coupling interfaces, so two meshes: `Left-Mesh` and `Right-Mesh`. How to distinguish both?

To dispel my doubts, could you please give a pseudo code example to explain what you mean by the following?
> Ideally, we could directly get groups of parameters, such as: {mesh, read/write, data name}

An alternative to extending the API (or the only solution?) is what we are trying to do with the [controller](https://github.com/precice/controller): we generate both the preCICE config and all adapter configs from a high-level description. This high-level description needs to know about the solvers topology (on which boundary do I want to couple to whom), the coupling physics. The coupling numerics can be auto-generated from _experience_.   

Btw, the controller repo also has a [list of examples](https://github.com/precice/controller/tree/master/examples).

**Comment by MakisH** [comment]
The participant name we definitely need. I also know that every participant can have multiple meshes and that it can read/write multiple data sets over one mesh.

What I have in mind is that a participant asks for what preCICE already knows about it. In pseudocode (with many inaccuracies):

```
interfaces = list<structure{ mesh_name, operation, data_name }>

interfaces = precice.get_interfaces(participant_name)
for each interface : interfaces
{
  if ( interface.operation == "write" )
  {
    cplDataWriters += cplDataWriter(mesh_name, data_name)
  }
  else
  {
    cplDataReader += cplDataReader(mesh_name, data_name)
  }
}
```

So, what I want to get from preCICE for each participant is:
```xml
<read-data name="Heat-Flux" mesh="Fluid-Mesh-Centers"/>
<write-data name="Temperature" mesh="Fluid-Mesh-Nodes"/>
```

**Important addition** here is that I assume we can guess the type of reader/writer from the name (e.g. temperature/pressure etc). I think this is what you want to say with
> In `Constants.hpp` we define default names for data types, such as Displacements, Forces, ... and also a default name for the preCICE config.

In https://github.com/precice/openfoam-adapter/issues/30#issuecomment-454808884 I want to also associate each piece of data with a type known to the adapter. This would not be immediately possible with the approach I propose here, so  ideas are welcome.

The controller can help, but the two solutions are not exclusive. In fact, the controller solves a different, larger problem and this solution does not affect it.

**Comment by uekerman** [comment]
Thanks for the explanation. I think I better understand your suggestion now. 
For the "coupling physics", we could find a solution I guess.
But wouldn't that solution still have the issue of how to associate such a pair `(mesh_name, data_name)` with a certain patch?

**Comment by MakisH** [comment]
Indeed, I forgot about that. I guess then that we cannot really simplify anything here without changing the XML file to e.g. add a unique identifier for each `read/write-data` node.

**Comment by uekerman** [comment]
Yes, extending the preCICE config could be an option. Good point. Let's do
a (hacked) prototype during the next coding days with OpenFOAM? Then, we
could really see all the implications.

**Comment by fsimonis** [comment]
Closing as not planned since we are focusing to autogenerate this information.
